### DISCUSSION TITLE
Batch multiple writeAtom operations

### DISCUSSION THREAD

**Discussion by yf-yang** [pmndrs/jotai#2840] [General] [answered]
https://codesandbox.io/p/sandbox/great-scott-x28y5r

Say there is an atom that depends on multiple atoms. When one of those dependencies are updated, its perfectly fine. However, when I need to update multiple atom "simultaneously", there exists an intermediate state where the downstream "get" is triggered with only one dependency atom is updated and the other is remained the same.

I know current behavior does not affect rendering. The use case is that I'd like to add some assertion in the dependent atom, but when two dependencies are updated simultaneously, there exists an intermediate state that breaks the assertion while the beginning/ending state still satisfy the assertion.

So, are there some api to achieve that? If not it's OK, I'll try to figure out other solutions.

**Comment by dmaskasky** [comment]
Related: https://github.com/pmndrs/jotai/pull/2827

**Reply by yf-yang** [reply]
Much better than I expect! Go for it 👊

**Reply by dmaskasky** [reply]
Its a WIP. But I think the approach is sound.

I think new behavior should be:
Atoms accessed with `get` (and all of their dependencies) are recomputed in place and unmarked from pending recompute. Their dependents (deep) are marked as pending recompute.

Atoms updated with `set` (and all their dependents) are marked as pending recompute.

All pending recompute atoms are recomputed in flushPending.

**Reply by yf-yang** [reply]
I realize one thing, the PR is different from what I expect.

Your PR is willing to solve that when an atom is set, its downstream atoms should only be computed once.

I'm expecting that when setting two or more atoms, their downstream atoms should be computed once. #2827 still not resolve this issue because `flushPending` will be called each time a dependency's set is called.

To achieve that, either some auto-batching mechanism is needed (I think React wraps all the event handlers to make sure event handlers are auto batched), or some explicit transaction-like batch function should be created.

**Reply by dai-shi** [reply, marked as answer]
It's our design decision not to introduce `batch` function. If we were to add it, it should be a third-party library in https://github.com/jotaijs.

**Reply by yf-yang** [reply]
Yep, seems it's not a good idea to do something like that, as people rarely need such a feature.

**Reply by dmaskasky** [reply]
We have sorely missed you 🥲

I need help with:
1. this - Recompute defers to flushPending where possible
2. jotai-scope reimplementation using `store.unstable_derive`
3. `get` in writeFn should not add dependencies (currently it does for atoms deeper in the tree)
4. stack overflow issues with read and write for large atom graphs (replace recursion with iteration)

The PR is still a WIP, but it is not necessary to recompute in-place unless something is read with `get`.

**Comment by dai-shi** [comment]
Hey @yf-yang, haven't heard from you for a while. How's everything? Would you still be able to help jotai-scope project?

**Reply by yf-yang** [reply]
Hi, I'm good! I've been checking out jotai-scope issues now and then, though not always promptly (and @dmaskasky responds pretty quick 👍), but there doesn't seem to be much to do. If there's anything I can help with, let me know. I saw that issue #2827 is also trying to tackle the complex jotai-scope scenario.

**Reply by yf-yang** [reply]
Oh I haven't check discord for several months 😲

**Reply by dai-shi** [reply]
Yeah, we use Discord for some discussions. It seems like you left my server. We can use GithHub discussions if you prefer.

**Reply by yf-yang** [reply]
Is it poimandres? I'm still there. I just downloaded a discord desktop app, that could be better.

**Reply by dai-shi** [reply]
It's another one: https://discord.gg/MrQdmzd

**Comment by dmaskasky** [comment]
Batch functionality added in https://github.com/pmndrs/jotai/pull/2827
